app/helpers/ExerciseConfig/Pipeline/Box/Boxes/JudgeBox.php

```diff
@@ -113,19 +109,32 @@ private function getJudgeBinaryAndArgs(): array {
 
     $judgeType = null;
     if ($this->hasInputPortValue(self::$JUDGE_TYPE_PORT_KEY)) {
-      $judgeType = $this->getInputPortValue(self::$JUDGE_TYPE_PORT_KEY)->getValue();
+      $judgeType = strtolower($this->getInputPortValue(self::$JUDGE_TYPE_PORT_KEY)->getValue());
     }
 
-    // judge type decision logic
-    if (empty($judgeType) || strtolower($judgeType) === self::$RECODEX_NORMAL_TYPE) {
-      return [ConfigParams::$JUDGES_DIR . self::$RECODEX_NORMAL_TYPE, []];
-    } else if (strtolower($judgeType) === self::$RECODEX_SHUFFLE_TYPE) {
-      return [ConfigParams::$JUDGES_DIR . self::$RECODEX_SHUFFLE_TYPE, []];
-    } else if (strtolower($judgeType) === self::$DIFF_TYPE) {
-      return [self::$DIFF_BINARY, []];
+    static $judgeTypes = null;
```

Comment:
Wouldn't it be better (readability-wise) to make this a class constant? It's not like lazy initialization saves us a considerable amount of time here...

Reply:
It cannot be a constant nor class static member since the initialization uses non-constant expressions.

Reply:
Which expressions are problematic? I have used array constants before (`const`). `ConfigParams::$JUDGES_DIR` could be made constant too.

Reply:
Variables (not even static ones like ConfigParams::$JUDGE_DIR) cannot be used in member variable (property) initialization nor in class constant initialization. Initialization expression must be constant (so it can be resolved at compile time).